light_curves/scale_up.md

```diff
@@ -0,0 +1,559 @@
+For example, other than ZTF's memory spike at the end, we see that the full run collecting multi-wavelength light curves for the SDSS 500k sample could be completed with about 2 CPU and 60G RAM.
+
+We want to learn why the ZTF job failed.
+Let's zoom in on that time period:
+
+```{code-cell}
+fig = toplog.plot_overview(between_time=("22:55", "23:10"))
+```
+
+In the second panel above we see the ZTF worker processes (which load the light curve data) ending just after 23:00 and then the ZTF parent process continues by itself.
+Around 23:06 in the fourth panel, we see the ZTF job's memory usage rise sharply to almost 100% and then drop immediately to zero when the job terminates.
+This coincides with the total available memory dropping to near zero in the third panel.
+This shows that the machine did not have enough memory for the ZTF call to successfully transform the light curve data collected by the workers into a `MultiIndexDFObject` and write it as a parquet file, so the machine killed the job.
+
+The solution is to rerun the ZTF archive call on a machine with more RAM.
+
+To learn exactly which step in `ztf_get_lightcurves` was causing this and how much memory it actually needed, additional print statements were inserted into the code similar to the following:
+
+```{code-cell}
+print(f"{helpers.scale_up._now()} | starting explode", flush=True)  # _now() prints the current timestamp
+```
+
+ZTF was then rerun on a large machine and `top` output was saved.
+After the run, we manually compared timestamps between the ZTF log and `top` output and tagged relevant `top` timestamps with corresponding step names by appending the name to the '----' delineator, like this for the "explode" step:
+
+```{code-cell}
+!cat output/lightcurves-demo-SDSS-500k/logs/top.tag-ztf.txt | grep -A12 explode
+```
+
+The helper can recognize these tags and show them on a figure:
+
+```{code-cell}
+ztf_toplog = helpers.top.load_top_output(toptxt_file="top.tag-ztf.txt", toptxt_dir=logs_dir, run_id=run_id)
+
+fig = ztf_toplog.plot_time_tags(summary_y="used_GiB")
+# (This run starts by reading a previously cached parquet file containing the raw data returned by workers.)
+```
+
+This figure shows that almost 100G RAM is required for the ZTF job to succeed.
+It further shows that the "explode" step requires the most memory, followed by creating the MultiIndexDFObject.
+From here, the user can choose an appropriately sized machine and/or consider whether `ztf_get_lightcurves` could be made to use less memory.
+
++++
+
+## Example 2: Parallelizing the light_curve_generator notebook
+
++++
+
+This example shows how to parallelize the example from the light_curve_generator notebook using the helper and python's `multiprocessing`.
+An advantage of the example shown here is that it automatically saves the sample and light curve data to disk after loading them and can automatically skip those steps in subsequent calls and use the files instead.
+
```

Comment:
Actually there's no difference in performance here vs the light_curve_generator parallel call. The only advantages are what I mention above about saving files and that this will print out the PIDs so you know which job is running in which process. I clarified by adding:

> This is a small sample (Yang, 30 objects). If you want a sample larger than a few hundred, consider using the bash script instead.